### ISSUE TITLE
The framework [.../onnxruntime.framework] is a framework of static libraries, and will not be copied to the app

### ISSUE THREAD

**Issue by rolfbjarne** [xamarin/xamarin-macios#21450]
_From @fdevillard on Mon, 14 Oct 2024 15:55:14 GMT_

### Describe the bug
I've created a toy MAUI project and everything build and run well until I add and use the following dependencies:
```
    <PackageReference Include="Microsoft.ML.OnnxRuntime" Version="1.19.2" />
    <PackageReference Include="Microsoft.ML.OnnxRuntime.Managed" Version="1.19.2" />
```

Whenever I try to use a resource from the dependency, the app crashes (exception below).

The build generates the following warning that I assume is directly related:
```
warning MT7091: The framework obj/Debug/net8.0-ios/iossimulator-arm64/onnxruntime.xcframework.zip/onnxruntime.xcframework/ios-arm64_x86_64-simulator/onnxruntime.framework is a framework of static libraries, and will not be copied to the app.
```

I don't understand why, and I'm running out of ideas to fix this.

Thank you in advance for any help 🙌

(Note: this was first briefly mentioned here https://github.com/dotnet/sdk/issues/36527)

### To Reproduce
I've created repo with the code: https://github.com/fdevillard/maui-onnx-crash

To reproduce:
- run the app
- click/press the "click me" button

Expected: "model loaded" is rendered in the button, showing the it has indeed been loaded

Behavior: the app crashes.

### Exceptions
```
Unhandled Exception:
System.TypeInitializationException: The type initializer for 'Microsoft.ML.OnnxRuntime.NativeMethods' threw an exception.

 ---> System.DllNotFoundException: __Internal
   at Microsoft.ML.OnnxRuntime.NativeMethods..cctor()
   --- End of inner exception stack trace ---

   at Microsoft.ML.OnnxRuntime.SessionOptions..ctor()
```

### Further technical details
- Include the output of `dotnet --info`: `8.0.403`
- The IDE: VSCode `1.94.2`
- macOS: 14.7
- iPhone simulator: version 16, iOS 18 

_Copied from original issue dotnet/sdk#44161_

**Comment by rolfbjarne** [comment]
_From @nagilson on Tue, 15 Oct 2024 20:08:10 GMT_

Not sure if this should be directed towards Maui or the Runtime team, but probably Maui?

**Comment by rolfbjarne** [comment]
_From @marcpopMSFT on Tue, 15 Oct 2024 20:15:00 GMT_

@Redth @jonathanpeppers should this go to the Maui team and which repo should we route it to? That MT error code I think is from the simulator, right?

**Comment by rolfbjarne** [comment]
_From @jonathanpeppers on Tue, 15 Oct 2024 20:41:15 GMT_

Do we have a way to move this to https://github.com/xamarin/xamarin-macios?

**Comment by rolfbjarne** [comment]
Can you try adding this package, and see if that fixes it?

```xml
<PackageReference Include="Microsoft.ML.OnnxRuntime.Extensions.Dummy" Version="0.10.0-beta" />
```

**Comment by fdevillard** [comment]
Thank you for your help, it's really appreciated 🙌

I've tried this dependency (at `0.10.0-beta`) as well as the latest version (`0.12.0`), and I still have the same symptoms / behavior.

**Comment by rolfbjarne** [comment]
I can reproduce a problem, but strangely it's not the problem you're seeing.

The problem I'm seeing is effectively a duplicate of #21294 - so hopefully this will be fixed once #21294 is fixed as well.

**Comment by BlueReptile2** [comment]
Hi everyone! I worked hard and found out a solution that worked with my .NET 8.0 MAUI, kind of like Microsoft.ML.OnnxRuntime.Extensions.Dummy. I will skip right to the tutorial since I usually hate to read too much..

**Step 1 (I extracted the package from zip):**
• Go to  Microsoft.ML.OnnxRuntime nuget package folder and find the zip
C:\Users\<username>\.nuget\packages\microsoft.ml.onnxruntime\<used version 1.21.0 in my case>\runtimes\ios\native\onnxruntime.xcframework.zip

• **[IMPORTANT]** Extract the folder with right click > Extract all (the result should be ...\runtimes\ios\native\onnxruntime.xcframework\onnxruntime.xcframework\ ios-arm64 & ios-arm64_x86_64-simulator). _I am not sure if it still works if you go ahead and copy paste the folder from the zip since it will only have the 1st folder onnxruntime.xcframework not the nested one_

• **[Optional]** I also deleted the folder that had "maccatalyst" since i don't use it

**Step 2 (I manually changed .targets file):**
• Find the .target file
"C:\Users\<username>\.nuget\packages\microsoft.ml.onnxruntime\<used version>\buildTransitive\net8.0-ios15.4\Microsoft.ML.OnnxRuntime.targets"

• Open the file and change NativeReference from
<NativeReference Include="$(MSBuildThisFileDirectory)..\..\runtimes\ios\native\onnxruntime.xcframework.zip">
to
<NativeReference Include="$(MSBuildThisFileDirectory)..\..\runtimes\ios\native\onnxruntime.xcframework">

• I also changed SmartLink from True to False, not sure if it makes any difference in this case.

**_AFTER THESE STEPS I TRIED TO RUN BUT HAD ANOTHER ISSUE:_**
_clang++ exited with code 1:
Undefined symbols for architecture arm64:
  "_RegisterCustomOps", referenced from:
      <initial-undefines>
ld: symbol(s) not found for architecture arm64
clang++: error: linker command failed with exit code 1 (use -v to see invocation).

**Step 3 [This step was completely made on a Mac then moved to the Visual Studio project] (I manually created a file to make the code work):**
• Open Terminal and put
`xcrun --sdk iphoneos --show-sdk-path`

• If that works, you are good to go ahead and write
`mkdir ~/Desktop/onnxstub`

Or create a folder manually on your Desktop

`cd ~/Desktop/onnxstub` (Go to the folder)

`SDK=$(xcrun --sdk iphoneos --show-sdk-path)` (Create a variable to use so it won't build for mac as i first did and got another error in Visual Studio)

`clang -arch arm64 -isysroot $SDK -c stub_register_custom_ops.c -o stub_register_custom_ops.o` (Create .o file)

`libtool -static -o libregister_custom_ops_stub.a stub_register_custom_ops.o` (Create .a file)

**Step 4 (I transferred and added the .a file to Visual Sudio project):**
• Use any way to bring the .a file to your Windows machine

• Place the .a file in Platforms/iOS/NativeStubs/ (or anywhere you want but make sure to change the path in next part)

• Add this snippet to the .csproj:
```
<NativeReference Include="Platforms/iOS/NativeStubs/libregister_custom_ops_stub.a">
  <Kind>Static</Kind>
  <ForceLoad>true</ForceLoad>
</NativeReference>
```

**Step 5 (The problem was persisting and this is what I did after, you might want to try if you get some of the problems):**

Problems:
- "Object not set to an instance of an object." (It sounds like a code problem but for me it was not)
- Again the error with _RegisterCustomOps appeared
- Some other errors that were crazy stuff

What I did:
• Delete bin and obj folders
• Restore Nuget Packages
• Clean Project
• Rebuild project
• Build on ios physical device using local device (this is to make sure the app builds and runs if the ONNX is not instantly used)
• Build over Remote Devices ([how to add a remote device](https://developer.apple.com/documentation/xcode/running-your-app-in-simulator-or-on-a-device#Connect-real-devices-to-your-Mac))

I guess you can also use a Simulator device but for me Simulator doesn't work.

After all the steps above my app started working. I hope this will help you all!
